src/movement/control_system.cpp

```diff
@@ -1,557 +1,573 @@
-     */
-    dt = 1.0/dt;
-
-    /*
-     * Initialize the state of the goals to error.
-     */
-    for(int i=0; i < 6; ++i)
-    {
-        goal_types[i] = robosub::control::STATE_ERROR;
-    }
-
-    /*
-     * Ensure that the commands and states are all set to zero initially.
-     */
-    state_vector = Vector12d::Zero();
-    current_integral = Vector6d::Zero();
-
-    /*
-     * Scale thruster limits to be in terms of the backwards thruster ratio.
-     */
-    t_lim *= back_thrust_ratio;
-    r_lim *= back_thrust_ratio;
+#include "control_system.h"
 
-    /*
-     * Load thruster node settings. A nodehandle reference is not available, so
-     * make use of rosparam get.
+namespace robosub
+{
+    /**
+     * Constructor.
      */
-    XmlRpc::XmlRpcValue thruster_settings;
-    if(!ros::param::get("thrusters", thruster_settings))
+    ControlSystem::ControlSystem()
     {
-        ROS_FATAL("Failed to load thruster parameters.");
-        exit(1);
-    }
-    ROS_INFO_STREAM("Loaded " << thruster_settings.size() << " thrusters.");
-    num_thrusters = thruster_settings.size();
+        ReloadPIDParams();
 
-    MatrixXd position = MatrixXd(num_thrusters,3);
-    MatrixXd orientation = MatrixXd(num_thrusters,3);
-    motors = MatrixXd(6,num_thrusters);
-    motor_commands = VectorXd::Zero(num_thrusters);
+        /*
+         * Load parameters from the settings file.
+         */
+        int rate = 0;
+        ros::param::getCached("/control/mass", sub_mass[0]);
+        ros::param::getCached("/control/mass", sub_mass[1]);
+        ros::param::getCached("/control/mass", sub_mass[2]);
+        ros::param::getCached("/control/inertia/psi", sub_mass[3]);
+        ros::param::getCached("/control/inertia/phi", sub_mass[4]);
+        ros::param::getCached("/control/inertia/theta", sub_mass[5]);
+        ros::param::getCached("/control/back_thrust_ratio", back_thrust_ratio);
+        ros::param::getCached("/control/limits/translation", t_lim);
+        ros::param::getCached("/control/limits/rotation", r_lim);
+        ros::param::getCached("/control/max_thrust", max_thrust);
+        ros::param::getCached("/control/rate", rate);
 
-    for(int i = 0; i < num_thrusters; ++i)
-    {
         /*
-         * Load in the thruster positions and orientations. The column of the
-         * position and orientation matrices denote the x, y, and z components
-         * sequentially.
+         * Calculate the change in time between each call.
+         * TODO: Replace dt with actual message timestamps.
          */
-        position(i,0) = thruster_settings[i]["position"]["x"];
-        position(i,1) = thruster_settings[i]["position"]["y"];
-        position(i,2) = thruster_settings[i]["position"]["z"];
-        orientation(i,0) = thruster_settings[i]["orientation"]["x"];
-        orientation(i,1) = thruster_settings[i]["orientation"]["y"];
-        orientation(i,2) = thruster_settings[i]["orientation"]["z"];
-    }
+        dt = 1.0/rate;
```

Comment:
Rate is loaded from the parameter server, which is listed as 20. We could have a ros error if the rate is specified as zero, but if this is in the constructor there isn't much we can do to handle that case. Regardless, currently we need a dt value for calculating the integral.